Simplify data export/import process

stognini opened the pull request:
This PR updates the Geant4 data export/import to a simpler mechanism:

- All exported/imported data is stored in a single struct named `ImportData`.
- Except for `ImportData`, all other import classes used by _geant-exporter_ were moved to `io/detail`.
- _geant-exporter_ stores the populated `ImportData` into a single TTree entry, vastly simplifying both writing and reading processes.
- `RootImporter::operator()("tree_name", "branch_name")` returns the `ImportData` object stored in the ROOT file. It now takes the names of the tree and branch as input parameters to avoid any hardcoded string name.
- All Celeritas' host/device classes now have a `from_import(ImportData& data)` function that returns a constructed `shared_ptr`.
- `test/gtest/detail/Macros.hh` now includes `TEST_IF_CELERITAS_USE_ROOT`. This allows disabling individual tests accordingly and is currently used to enable/disable all `from_import(...)` tests added to the host/device classes.

Example code on how the data is now imported into Celeritas:
```
RootImporter import("/path/to/root_file.root");
const auto   data            = import("tree_name", "branch_name");
const auto   particle_params = ParticleParams::from_import(data);
```

sethrj commented on test/io/GdmlGeometryMap.test.cc:
```diff
@@ -0,0 +1,85 @@
+        root_filename_ = this->test_data_path("io", "geant-exporter-data.root");
+        RootImporter import_from_root(root_filename_.c_str());
+        data_ = import_from_root("geant4_data", "ImportData");
+        ;
```
Typo that got clang-formatted :)

sethrj commented on test/physics/base/Particle.test.cc:
```diff
@@ -80,6 +85,59 @@ TEST_F(ParticleTest, params_accessors)
+        root_filename_ = this->test_data_path("io", "geant-exporter-data.root");
+        RootImporter import_from_root(root_filename_.c_str());
+        data_ = import_from_root("geant4_data", "ImportData");
+        ;
```
Another typo.

sethrj commented on test/physics/base/Particle.test.cc:
```diff
@@ -80,6 +85,59 @@ TEST_F(ParticleTest, params_accessors)
+        loaded_pdgs.push_back(particles->id_to_pdg(particle_id).get());
+    }
+
+    // Particle ordering is the same as in the ROOT file
```
This comment is no longer accurate, right? The ROOT file now stores them in Geant4's internal ordering, and the `ParticleParams::from_import` reorders them.

sethrj commented on test/physics/em/LivermorePE.test.cc:
```diff
@@ -17,7 +17,7 @@
 #include "comm/Device.hh"
 #include "io/AtomicRelaxationReader.hh"
-#include "io/ImportPhysicsTable.hh"
+#include "io/detail//ImportPhysicsTable.hh"
```
Typo with the double-slash—but this should be reverted once these `Import` classes get moved back out of detail.

sethrj commented on test/physics/material/Material.test.cc:
```diff
@@ -227,6 +230,72 @@ TEST_F(MaterialTest, element_view)
+        root_filename_ = this->test_data_path("io", "geant-exporter-data.root");
+        RootImporter import_from_root(root_filename_.c_str());
+        data_ = import_from_root("geant4_data", "ImportData");
+        ;
```
Same typo

sethrj reviewed (changes requested):
Great! This is on the right track and probably 95% to the goal. The `RootImporter` operator is a little funky, and some of the `from_import` methods need to be handed existing "pre-converted" data structures rather than rebuilding them.

stognini commented:
Now that `ImportData` already includes all the needed information to connect volumes and materials, the `GdmlGeometryMap` class is not fundamentally needed anymore. Nevertheless, it provides an easy interface to read the material, element, and volume information while constructing new classes. It is now constructed from a populated `ImportData` object, as seen in `GdmlGeometryMap.test`, `GeoMaterial.test`, and `LDemoParams`.

Two points of discussion are: 
- Do we want to keep it, since it provides a better interface to find materials, elements, and volumes? Or should we just delete it and read the vectors available in `ImportData` directly when constructing other objects?
- If we intend to keep it, we might revisit the `GdmlGeometryMap` name now that it serves as an interface.

sethrj commented:
> Now that `ImportData` already includes all the needed information to connect volumes and materials, the `GdmlGeometryMap` class is not fundamentally needed anymore. Nevertheless, it provides an easy interface to read the material, element, and volume information while constructing new classes. It is now constructed from a populated `ImportData` object, as seen in `GdmlGeometryMap.test`, `GeoMaterial.test`, and `LDemoParams`.
> 
> Two points of discussion are:
> 
> * Do we want to keep it, since it provides a better interface to find materials, elements, and volumes? Or should we just delete it and read the vectors available in `ImportData` directly when constructing other objects?
> * If we intend to keep it, we might revisit the `GdmlGeometryMap` name now that it serves as an interface.

I'd like the "import" layer to be as simple as possible, so we move the data as quickly as possible into our more feature-rich Params classes, which offers the finding/mapping functionality we need. It could make sense to have an extra mapping layer at the import layer if we were filtering out data at problem setup time, but at this point it looks like our import layer will always be loading *all* of the data because it's so interconnected. Any extra useful mapping functionality should be added to the Params classes or a helper function/class that uses multiple params classes.

I don't *think* that removing the GdmlMap would make construction more complicated—the data is stored as vectors and we need it as vectors, so there shouldn't be much mapping involved.

stognini commented:
> I don't think that removing the GdmlMap would make construction more complicated—the data is stored as vectors and we need it as vectors, so there shouldn't be much mapping involved.

It mostly does not, so I have removed `GdmlGeometryMap` in its entirety. I have just added 2 small free functions in `ImportData` to make life a little bit easier. They retrieve full elemental or material data from a given `element_id` or `material_id`. Element, material, and volume IDs are defined as `unsigned int` to simplify this import layer as much as possible and not rely on any particular type, but let me know if you believe that a type-safe index is fundamental.

sethrj commented on app/geant-exporter/ImportProcessConverter.cc:
```diff
@@ -302,14 +290,27 @@ void GeantPhysicsTableWriter::operator()(const G4ParticleDefinition& particle,
+/*!
+ * Remove any empty ImportProcess returned by operator().
+ */
+void ImportProcessConverter::remove_empty(std::vector<ImportProcess>& processes)
```
Since this is only used in the one place, I think it would be better to move to the calling location. If you disagree (I understand that this idiom is a little ugly so it's nicer not to have to think about it), you should either:
- use `*` instead of mutable `&`, or
- take a `vector&&` and return a `vector` by value, so that the caller would do `vec = remove_empty(std::move(vec))`

and also make this method `static` since it doesn't require class data.

sethrj commented on src/io/ImportData.cc:
```diff
@@ -0,0 +1,58 @@
+ * This simplifies the process of fetching full elemental information from a
+ * given material, as \c ImportMatElemComponent only stores the \c element_id .
+ */
+const ImportElement
```
```suggestion
const ImportElement&
```

sethrj commented on src/io/ImportElement.hh:
```diff
@@ -18,21 +18,22 @@ namespace celeritas
-    real_type   atomic_mass;           // [atomic mass unit]
-    real_type   radiation_length_tsai; // [g/cm^2]
-    real_type   coulomb_factor;
+    unsigned int element_id;
```
Why is this new field necessary? Isn't it just the index of this element in the vector of `ImportElement`s?

sethrj commented on src/io/ImportMaterial.hh:
```diff
@@ -29,39 +29,52 @@ enum class ImportMaterialState
-    std::map<int, ProductionCut> pdg_cutoff;             // [MeV, cm]
-    std::map<elem_id, real_type> elements_fractions;     // Mass fractions
-    std::map<elem_id, real_type> elements_num_fractions; // Number fractions
+    unsigned int                        material_id;
```
This is also kind of unnecessary unless you somehow have a pointer to an `ImportMaterial` rather than a whole vector of them.

sethrj commented on src/io/ImportData.cc:
```diff
@@ -0,0 +1,58 @@
+                             data.elements.end(),
+                             [element_id](const ImportElement& element) {
+                                 return element.element_id == element_id;
+                             });
```
Shouldn't this just be `auto iter = data.elements.begin() + element_id`? The "id" of an element should always be its position in the list, if we're being consistent and operating on import data in the context of an `ImportFile`.

sethrj reviewed (changes requested):
Looks great, Stefano, thanks for the updates. Regarding the helper functions: are they used anywhere? I don't think we should ever need to "search" through the *import* data (especially since the search criterion is just the index of an element/material in its vector), and as I commented there I think the new element/material ID accessors are redundant in the context of loading an entire ImportFile.

stognini commented on app/geant-exporter/ImportProcessConverter.cc:
```diff
@@ -302,14 +290,27 @@ void GeantPhysicsTableWriter::operator()(const G4ParticleDefinition& particle,
+    processes.erase(
+        std::remove_if(processes.begin(),
+                       processes.end(),
+                       [](ImportProcess& ip) { return ip ? false : true; }),
```
`ImportProcess` already has a boolean operator that verifies if the struct is populated. I wrote it that way to be clearer to notice that if we have a populated `ImportProcess`, then return false (i.e. do not erase). If I just return bool ip, I invert that logic.

sethrj replied:
Oh whoops, I didn't even notice that—my bad. You should then write (even easier!)

```suggestion
                       [](ImportProcess& ip) { return !ip; }),
```

stognini commented:
Sorry for not waiting on another PR to address some new improvements, but they fit in the current scope. After we moved from a "geometry" idea (`GdmlGeometryMap` mapped elements, materials, and volumes) to a struct of vectors in `ImportData`, I refactored the whole code for consistency, better readability, and maintenance. So now both _geant-exporter_ and _geant-exporter-cat_ have individual functions to write/print each member of `ImportData`:

- geant-exporter:
```
// Populate ImportData members
import_data.particles = store_particles();
import_data.elements  = store_elements();
import_data.materials = store_materials();
import_data.processes = store_processes();
import_data.volumes   = store_volumes(world_phys_volume);
```
- geant-exporter-cat:
```
print_particles(*particle_params);
print_elements(data.elements);
print_materials(data.materials, data.elements, *particle_params);
print_processes(data.processes, *particle_params);
print_volumes(data.volumes, data.materials);
```

Currently _geant-exporter-cat_ print functions still rely on a mixed bag of `ImportData` and `Params`. This should be revisited soon.

sethrj commented on app/geant-exporter/geant-exporter.cc:
```diff
@@ -255,48 +109,153 @@ int to_pdg(const G4ProductionCutsIndex& index)
+    {
+        loop_volumes(volumes,
+                     *logical_volume.GetDaughter(i)->GetLogicalVolume());
+    }
```
This loop looks dangerous from a scaling perspective: every volume loops recursively into all its daughters, so that means this could be an `O(N^2)` operation. Is it possible to loop directly over logical volume IDs rather than hierarchically through the geometry? That would allow you to eliminate the `get_max_instance_id` function and juts directly loop over volumes. If not, it might be better here to temporarily use a map (or unordered map) of volume IDs -> volumes, and then convert to a vector by looping over indices.

stognini replied:
Yes. This loop has been debated between @mrguilima and me. It is a very inefficient loop that I am still not sure how to tackle, and I've been playing around with it since I pushed these commits. The `loop_volumes` ends up going over all geometry nodes (which represents a total of 2,104,795 nodes for the case of cms2018.gdml), because there is no way to loop over instances, as you said. There is no `GetInstanceID(G4int id)` or `GetLogicalVolume(G4int instance_id)`, so the only obvious way was to deal with it the hard way.

sethrj replied:
Thumbs down for the code, not for you :) you still might be able to recurse only once if you use a map rather than a vector

stognini replied:
At the end, looping over all nodes seems to be inevitable (I was hoping that some loops over daughter volumes could be skipped). So I accepted defeat and went for the map solution you mentioned. At least we loop only once over them to build the map and use it to populate the final `vector<ImportVolume>`.

sethrj commented on app/geant-exporter/geant-exporter.cc:
```diff
@@ -334,55 +293,128 @@ void store_geometry(TFile*                       root_file,
-        // Add material to the global material map
-        geometry.add_material(g4material_cuts_couple->GetIndex(), material);
+        // Add material to vector
+        materials[g4material_cuts_couple->GetIndex()] = material;
```
Also here it would be good to have a range check.

sethrj reviewed (commented):
Looks great aside from one final scaling issue in the geant exporter that I'd like to have addressed... sorry!

stognini commented on app/geant-exporter/geant-exporter.cc:
```diff
@@ -255,48 +109,153 @@ int to_pdg(const G4ProductionCutsIndex& index)
-    TBranch* branch = tree_materials.Branch("GdmlGeometryMap", &geometry);
-    CELER_ASSERT(branch);
+    // Volume position represents the volume id
+    volumes[logical_volume.GetInstanceID()] = volume;
```
Sure, although I believe this already could not happen, as I already have to find that size manually (there is no such thing as a `GetMaxInstanceID()` function in G4).

stognini replied:
PS: forgot to answer your question: it was not. Before it was a simple `push_back`, which could have led to wrong volume_ids. There is no guarantee that the `loop_volumes` will go over instance ids in a monotonically increasing way (so we could end up with a vector where an `import_volume.at(10)` represented an `instance_id == 3`). So I had to resize the vector and store volumes at the correct position manually.

sethrj commented on app/geant-exporter/geant-exporter.cc:
```diff
@@ -334,55 +276,133 @@ void store_geometry(TFile*                       root_file,
+    loop_volumes(volids_volumes, *world_volume->GetLogicalVolume());
+
+    // Populate vector<ImportVolume>
+    for (const auto& key : volids_volumes)
+    {
+        volumes.push_back(key.second);
```
One final change: if someone unwittingly replaces `map` with `unordered_map`, or if the logical volumes are discontiguous/offset for some reason, the loop will proceed as normal and the IDs  in the export file will be non-obviously mixed up.
```suggestion
    for (const auto& key : volids_volumes)
    {
        CELER_ASSERT(key.first == volumes.size());
        volumes.push_back(key.second);
```

stognini commented:
I've added a much simpler (and kinda obvious) solution regarding the empty `ImportProcess` objects returned by `ImportProcessConverter::operator()`. It makes the code cleaner and allows us to remove the `ImportProcessConverter::remove_empty()` function, which you were not very fond of.

We simply use `ImportProcess::operator bool()` to check if `ImportProcess` is empty before adding it to the vector:

```
if (ImportProcess process
    = process_writer(g4_particle_def, *process_list[j]))
{
        // Not an empty process, so it was not added in a previous loop
        processes.push_back(process);
}
```

sethrj commented:
@stognini great idea! One final gotcha there though is the distinction between copy and move. The `push_back(process)` does a recursive copy because `process` is an "lvalue".  Change to `push_back(std::move(process))` for the desired effect.

sethrj reviewed (approved):
Thanks a bunch,  @stognini !
